src/main/scala/mesosphere/marathon/storage/migration/Helper.scala

```diff
@@ -111,4 +116,63 @@ object InstanceMigration extends MaybeStore with StrictLogging {
       case ZkSerialized(byteString) =>
         Json.parse(byteString.utf8String)
     }
+
+  /**
+    * JSON [[Reads]] format for reservations that have no id saved. See [[MigrationTo17]] for an example.
+    *
+    * @param tasksMap The tasks from the instance belonging to the reservation.
+    * @param instanceId The id of the instance.
+    * @return The reservation with an inferred id or the already persisted id.
+    */
+  def legacyReservationReads(tasksMap: Map[Task.Id, Task], instanceId: Id): Reads[Reservation] = {
+    (
+      (__ \ "volumeIds").read[Seq[LocalVolumeId]] ~
+      (__ \ "state").read[Reservation.State] ~
+      (__ \ "id").readNullable[Reservation.Id]
+    ) { (volumesIds, state, maybeId) =>
+        val reservationId = maybeId.getOrElse(inferReservationId(tasksMap, instanceId))
+        Reservation(volumesIds, state, reservationId)
+      }
+  }
+
+  /**
+    * Infer the reservation id for an instance.
+    *
+    * In older Marathon versions the reservation id was the run spec path and the uuid of the instance
+    * joined by a separator. Eg ephemeral tasks and persistent apps used it. This is expressed by
+    * [[mesosphere.marathon.core.instance.Reservation.LegacyId]].
+    *
+    * Later Marathon versions used the instance id, ie `<run spec path>.marathon-<uuid>`, as the
+    * reservation id. This is expressed by [[mesosphere.marathon.core.instance.Reservation.SimplifiedId]].
+    * Notice the extra "marathon-" in the id string.
+    *
+    * The reservation id was in all cases determined by the `appTask.taskId`. The app task is just the
+    * first task of the tasks map of an instance.
+    *
+    * This method is only used if the saved reservation does not include an id. This will be true for
+    * all instance from apps and pods started with Marathon 1.8.194-1590825ea and earlier. All apps
+    * and pods from later version will have a reservation id persisted.
+    *
+    * Future Marathon versions that only allow upgrades from storage version (1, 8, 200) and later can
+    * drop the inference and should safely assume that all reservation have a persisted id.
+    *
+    * @param tasksMap All tasks of an instance.
+    * @param instanceId The id of the instance this reservation belongs to.
+    * @return The proper reservation id.
+    */
+  private[migration] def inferReservationId(tasksMap: Map[Task.Id, Task], instanceId: Id): Reservation.Id = {
+    if (tasksMap.nonEmpty) {
+      val taskId = appTask(tasksMap).getOrElse(throw new IllegalStateException(s"No task in $instanceId")).taskId
```

Comment:
can this fail on 1.8? are we still guaranteed to have task around?

Reply:
`appTask` works on `tasksMap`. Hence the `nonEmpty` check above. When can have empty task maps on 1.8.